### DISCUSSION TITLE
[General] Gas Limit Computation

### DISCUSSION THREAD

**Discussion by sanjukammath** [zkSync-Community-Hub/zksync-developers#191] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

This logic to compute gas limit externally is required when we are estimating a transaction that will fail on the chain because of a dependent transaction not being send.
For example swapping in a dex requires an approval transaction. Without approval transaction swap transaction will fail. Also the eth_estimateGas will also throw unpredictable gas limit error. But we would need to estimate the swap before submitting approval. 

### What is the specific issue or error you're encountering?

We normally compute such gas limits by simulating the submitting of approval on a fork. However if we use a fork using hardhat, we end up getting the regular EVM gas limit. If we have a formula to take that gas limit and find out ZKSYNC gas limit it will be helpful.

### Can you share the error messages or logs you're receiving, if any?

NA

### Have you made any recent changes to the contract before encountering this issue?

NA

### Are there any external libraries or contracts that your contract interacts with?

NA

### Can you provide the relevant portions of your contract code where the issue is occurring?

NA

### Have you tried to isolate the problem, and if so, what were the results?

NA

### What steps have you already taken to try to resolve the issue?

NA

### Additional Details

Other L2 chains like arbitrum provide a formula to identify how the gas limit on their chains can be derived from the gas limit on the L1 chain like so:
https://docs.arbitrum.io/devs-how-tos/how-to-estimate-gas#breaking-down-the-formula

**Comment by MexicanAce** [comment, marked as answer]
Hi @sanjukammath 👋 !

We have the following document to help outline the cost of gas fees for zkSync Era : https://github.com/matter-labs/zksync-era/blob/main/docs/advanced/gas_and_fees.md

There's not exact formula to convert L1 gas limit to zkSync gas limit, as that requires a full estimation. This is due to an overhead and cost savings implemented when batching similar transactions that access similar storage slots.

Hopefully this answers your question, let us know if there's anything we can help answer.

**Comment by kimberleeni** [comment]
It would be beneficial to implement a formula to compute the gas limit for zkSync transactions based on the gas limit of the corresponding Ethereum transactions. This would help estimate gas accurately for zkSync transactions that depend on Ethereum transactions, such as approval transactions in decentralized exchanges. Similar to Arbitrum's approach, having a formula tailored to zkSync would streamline transaction estimation and enhance developer experience.

**Comment by etesvipatdeb** [comment]
Sure! We'll explore options for deriving ZKSync gas limits from the Ethereum gas limit to facilitate estimating gas for transactions. Thank you for bringing this to our attention!

**Reply by sanjukammath** [reply]
It will be fine if you can support state overrides for the eth_estimateGas RPC call. This is already available in chains like Ethereum BSC and Arbitrum. This will help us override the approval value (or any required state) and make an estimate call.

**Comment by sfatitetyccindy** [comment]
Issue: Need to compute gas limit for zkSync transactions based on Ethereum gas limit.
Possible Solution: Develop a formula to derive zkSync gas limit from Ethereum gas limit.
Next Steps: Research zkSync documentation or contact zkSync developers for guidance on gas limit computation.

**Comment by epjocnopasara** [comment]
It seems you're looking for a method to compute gas limits for transactions on zkSync based on gas limits from the Ethereum mainnet. This functionality would facilitate estimating gas for transactions that involve multiple steps or dependent transactions. Having a formula similar to what Arbitrum provides could help developers better estimate gas costs for zkSync transactions.

**Comment by venlaudecsubs** [comment]
The user is trying to implement logic to compute gas limit externally on the zkSync testnet. This is required when estimating a transaction that will fail on the chain due to a dependent transaction not being sent. The user normally computes such gas limits by simulating the submitting of approval on a fork, but when using hardhat, they end up getting the regular EVM gas limit. The user is looking for a formula to take the regular EVM gas limit and find out the zkSync gas limit. Other L2 chains like Arbitrum provide a similar formula for identifying how the gas limit on their chains can be derived from the gas limit on the L1 chain. No specific error messages or logs are provided, and no recent changes have been made to the contract. There are no external libraries or contracts that the contract interacts with, and no relevant portions of the contract code are provided. No steps have been taken to isolate the problem, and no repo link is provided.
